### ISSUE TITLE
items.grep with a RegEx should return items whose names match the regex

### ISSUE THREAD

**Issue by boc-tothefuture** [boc-tothefuture/openhab-jruby#424]
Right now `items.grep` with a regex only works with Strings and searches across the values of those String items.  The proposal here is a breaking change and makes items.grep with a regex match on the item names for all item types.

@jimtng @ccutrer ?

**Comment by jimtng** [comment]
> Right now `items.grep` with a regex only works with Strings and searches across the values of those String items.

How would we search across the values of items after changing grep to match on item names?

**Comment by ccutrer** [comment]
What's the implementation?

**Comment by boc-tothefuture** [comment]
@ccutrer Not sure on the implementation, I have to refresh myself with how grep and === interact. 

@jimtng I think only regex would be impacted, not the class checks or the number checks.

**Comment by ccutrer** [comment]
this is essentially it's definition:
```
module Enumerable
  def grep(v)
    select { |i| v === i }
  end
end
```

which means you need to think about `Regexp#===`. `Regexp#===` is essentially defined as:
```
class Regexp
  def ===(other)
    other = if other.instance_of?(Symbol)
        other.to_s
      elsif other.instance_of?(String)
        other
      elsif other.respond_to?(:to_str)
        other.to_str
      else
        raise TypeError
      end
    return false if other.nil?
    match?(other)
  end
end
```

So you need to inject in at the `GenericItem#to_str` call. The problem is that `StringItem#to_str` is already in use, to implicitly convert to the _state_ of the item for various reasons. And I would _highly_ recommend against implementing `#to_str` for other item types, as lots of core ruby stuff will suddenly start calling it and have surprising results.

The other option would be to override `Regexp#===` to introduce custom logic. The problem with doing that is it's used for case statements. So you will break constructs like
```
case MyStringItem
when /some_regex/
end
```
to suddenly start matching against the item's name instead of the item's state.

Tl;dr: I'm fairly certain this is not possible without breaking other things. I would suggest either `items.select { |i| i.name.match?(/regex/) }` isn't really that onerous, or introducing a custom method `items.grep_by_name(/regex/)` that just does the select internally as above.

**Comment by jimtng** [comment]
grep_by_name sounds nice and less confusing

**Comment by boc-tothefuture** [comment]
Closing based on feedback.
